### PR TITLE
NP-1441 Split Script run/event into own DAO

### PR DESCRIPTION
**Pull request by jlhughes**
Refactoring which occurred during Medusa development. 
Also correct an interaction between StorageOptional and Cronjobs which prevented the cronjob scheduledTime updates to not be stored in the MDAO.

### CODE REVIEW

File: src/foam/nanos/boot/Boot.java [hunk 1]
```diff
@@ -41,7 +41,7 @@ public Boot(String datadir) {
     XLocator.set(root_);
 
     Logger logger = new ProxyLogger(new StdoutLogger());
-    root_.put("logger", logger);
+    root_ = root_.put("logger", logger);
```
Comment (kgrgreer) [inline src/foam/nanos/boot/Boot.java]: Why is this necessary? root_ is a ProxyX, which does this automatically.

File: src/foam/nanos/boot/Boot.java [hunk 2]
```diff
@@ -113,17 +117,35 @@ public Object create(X x) {
...
       Script script    = (Script) scriptDAO.find(startScript);
       if ( script != null ) {
-        ((Script) script.fclone()).runScript(root_);
+        logger.info("Boot,script", startScript);
+        script = (Script) script.fclone();
+        // REVIEW: want main to run in ReadOnlyContext, but tests to run normally
+        if ( "main".equals(startScript) ) {
+          try {
+            // NOTE: is read-only and will throw exception when it updates rundate.
+            script.runScript(new foam.core.ReadOnlyDAOContext(root_));
```
Comment (kgrgreer) [inline src/foam/nanos/boot/Boot.java]: Why is it ReadOnly?

File: src/foam/nanos/cron/CronScheduler.java [hunk 3]
```diff
@@ -55,30 +67,34 @@ public void run() {
...
-          }
-        });
+        if ( getEnabled() ) {
+          Date now = new Date();
 
+          cronDAO_.where(
+                         MLang.AND(
+                                   MLang.LTE(Cron.SCHEDULED_TIME, now),
+                                   MLang.EQ(Cron.ENABLED, true),
+                                   MLang.OR(
```
Comment (kgrgreer) [inline src/foam/nanos/cron/CronScheduler.java]: IN would be easier / more efficient here.

File: src/foam/nanos/cron/CronScheduler.java [hunk 4]
```diff
@@ -55,30 +67,34 @@ public void run() {
...
+                                            MLang.EQ(Cron.STATUS, ScriptStatus.UNSCHEDULED),
+                                            MLang.EQ(Cron.STATUS, ScriptStatus.ERROR)
+                                            )
+                                   )
+                         ).select(new AbstractSink() {
```
Comment (kgrgreer) [inline src/foam/nanos/cron/CronScheduler.java]: Should there also be an orderBy() so they're run in the scheduled order?

File: src/foam/nanos/cron/CronScheduler.java [hunk 5]
```diff
@@ -89,9 +105,13 @@ public void put(Object obj, Detachable sub) {
           if ( delay > 5000 ) {
```
Comment (kgrgreer) [inline src/foam/nanos/cron/CronScheduler.java]: Code simpler if this is just: delay > 5000 is ! getEnabled()

File: src/foam/nanos/boot/Boot.java (hunk 2 above)
Comment (jlhughes) [reply src/foam/nanos/boot/Boot.java]: I don't want it to save back into the journals/bootscript file.  The bootscript is just that, boot only, can't have side effects for clustering.

File: src/foam/nanos/cron/CronScheduler.java (hunk 4 above)
Comment (jlhughes) [reply src/foam/nanos/cron/CronScheduler.java]: good idea.  Mind you, on 'put' the ScriptRunnerDAO calls agency.execute so order is lost.

File: src/foam/nanos/boot/Boot.java (hunk 1 above)
Comment (jlhughes) [reply src/foam/nanos/boot/Boot.java]: updated.

File: src/foam/nanos/cron/CronScheduler.java (hunk 3 above)
Comment (jlhughes) [reply src/foam/nanos/cron/CronScheduler.java]: changed.

File: src/foam/nanos/cron/CronScheduler.java (hunk 4 above)
Comment (jlhughes) [reply src/foam/nanos/cron/CronScheduler.java]: added

File: src/foam/nanos/boot/Boot.java (hunk 2 above)
Comment (kgrgreer) [reply src/foam/nanos/boot/Boot.java]: But we have other bootscripts when we want to run unit tests from the command-line, like the "TestReporter" script. Will this break scripts like those?

Comment (jlhughes) [reply src/foam/nanos/boot/Boot.java]: no, only 'main' is run in a ReadOnlyDAOContext.

Comment (jlhughes) [reply src/foam/nanos/boot/Boot.java]: Removed. 
This readonly restriction on main has been in the medusa branch a long time. Whatever it was protecting is no longer an issue.
